addons/sale_timesheet/tests/test_reinvoice.py

```diff
@@ -283,3 +287,92 @@ def test_no_expense(self):
+            'standard_price': 30,
+            'list_price': 90,
+            'type': 'service',
+            'service_policy': 'delivered_timesheet',
+            'invoice_policy': 'delivery',
+            'default_code': 'SERV-DELI2',
+            'service_type': 'timesheet',
+            'service_tracking': 'task_global_project',
+            'project_id': self.project_global.id,
+            'taxes_id': False,
+            'property_account_income_id': self.account_sale.id,
+        })
+        today = Date.context_today(self.env.user)
+
+        # Creates a sales order for quantity 3
+        so_form = Form(self.env['sale.order'])
+        so_form.partner_id = self.env['res.partner'].create({'name': 'Toto'})
+        with so_form.order_line.new() as line:
+            line.product_id = product
+            line.product_uom_qty = 3.0
+        sale_order = so_form.save()
+        sale_order.action_confirm()
+
+        # "Deliver" 1 of 3
+        task = sale_order.tasks_ids
+        self.env['account.analytic.line'].create({
+            'name': 'Test Line',
+            'project_id': task.project_id.id,
+            'task_id': task.id,
+            'unit_amount': 1,
+            'employee_id': self.employee_user.id,
+            'company_id': self.company_data['company'].id,
+        })
+
+        context = {
+            "active_model": 'sale.order',
+            "active_ids": [sale_order.id],
+            "active_id": sale_order.id,
+            'open_invoices': True,
+        }
+        # Invoice the 1
+        wizard = self.env['sale.advance.payment.inv'].with_context(context).create({
+            'advance_payment_method': 'delivered'
+        })
+        invoice_dict = wizard.create_invoices()
+        # Confirm the invoice
+        invoice = self.env['account.move'].browse(invoice_dict['res_id'])
+        invoice.action_post()
+        # Refund the invoice
+        refund_invoice_wiz = self.env['account.move.reversal'].with_context(active_model="account.move", active_ids=[invoice.id]).create({
+            'reason': 'please reverse :c',
+            'refund_method': 'refund',
+            'date': today,
+        })
+        refund_invoice = self.env['account.move'].browse(refund_invoice_wiz.reverse_moves()['res_id'])
+        refund_invoice.action_post()
+        # reversing with action_reverse and then action_post does not reset the invoice_status to 'to invoice' in tests
+
+        # Recreate wizard to get the new invoices created
+        wizard = self.env['sale.advance.payment.inv'].with_context(context).create({
+            'advance_payment_method': 'delivered',
+            'date_start_invoice_timesheet': today - timedelta(days=7),
+            'date_end_invoice_timesheet': today + timedelta(days=7)
+        })
+
+        try:
+            wizard.create_invoices()
+        except UserError:
+            raise Exception("We reversed the journal entry, so there is still quantity 3 to invoice")
```

Comment:
I am not sure this is clear. Why not just calling wizard.create_invoices() and checking that a new invoice has been generated with a quantity of 1?

Reply:
Because i'm trying to simulate the issue in the business flow, everything is fine until we try to create an invoice, where a UserError saying "There is nothing to invoice!" is raised there, but that shouldn't happen after my commit

We could call ourselves `_recompute_qty_to_invoice` to have https://github.com/odoo/odoo/blob/94e21b570e36e36d9b58edf84e10e4a7bb46581b/addons/sale_timesheet/models/sale_order.py#L232 called, but it made more sense to me in the test to follow the use case and ensure there is no error at that time

I would have rather put except ThereIsNothingToInvoice or something similar but the exception raised is just UserExcept :/

Reply:
For me the description of the UserError is correct so I would rather not encapsulate it in the try catch.